lib/Conversion/HandshakeToDC/HandshakeToDC.cpp

```diff
@@ -0,0 +1,612 @@
+                             index, hw::InnerSymAttr{}});
+  }
+
+  return ports;
+}
+
+class FuncOpConversion : public DCOpConversionPattern<handshake::FuncOp> {
+public:
+  using DCOpConversionPattern<handshake::FuncOp>::DCOpConversionPattern;
+  using OpAdaptor =
+      typename DCOpConversionPattern<handshake::FuncOp>::OpAdaptor;
+
+  // Replaces a handshake.func with a hw.module, converting the argument and
+  // result types using the provided type converter.
+  // @mortbopet: Not a fan of converting to hw here seeing as we don't
+  // necessarily have hardware semantics here. But, DC doesn't define a function
+  // operation, and there is no "func.graph_func" or any other generic function
+  // operation which is a graph region...
+  LogicalResult
+  matchAndRewrite(handshake::FuncOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    ModulePortInfo ports = getModulePortInfo(*getTypeConverter(), op);
+
+    if (op.isExternal()) {
+      rewriter.create<hw::HWModuleExternOp>(
+          op.getLoc(), rewriter.getStringAttr(op.getName()), ports);
+    } else {
+      auto hwModule = rewriter.create<hw::HWModuleOp>(
+          op.getLoc(), rewriter.getStringAttr(op.getName()), ports);
+
+      auto &region = op->getRegions().front();
+
+      Region &moduleRegion = hwModule->getRegions().front();
+      rewriter.mergeBlocks(&region.getBlocks().front(), hwModule.getBodyBlock(),
+                           hwModule.getBodyBlock()->getArguments());
+      TypeConverter::SignatureConversion result(moduleRegion.getNumArguments());
+      (void)getTypeConverter()->convertSignatureArgs(
+          TypeRange(moduleRegion.getArgumentTypes()), result);
+      rewriter.applySignatureConversion(&moduleRegion, result);
+    }
+
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+class HandshakeToDCPass : public HandshakeToDCBase<HandshakeToDCPass> {
+public:
+  void runOnOperation() override {
+    mlir::ModuleOp mod = getOperation();
+
+    // Maintain the set of operations which has been converted either through
+    // unit rate conversion, or as part of other conversions.
+    ConvertedOps convertedOps;
+
+    ConversionTarget target(getContext());
+    target.addIllegalDialect<handshake::HandshakeDialect>();
+    target.addLegalDialect<dc::DCDialect, func::FuncDialect, hw::HWDialect>();
+    target.addLegalOp<mlir::ModuleOp>();
+
+    // The various patterns will insert new operations into the module to
+    // facilitate the conversion - however, these operations must be
+    // distinguishable from already converted operations (which may be of the
+    // same type as the newly inserted operations). To do this, we mark all
+    // operations which have been converted as legal, and all other operations
+    // as illegal.
+    target.markUnknownOpDynamicallyLegal(
+        [&](Operation *op) { return convertedOps.contains(op); });
```

Comment:
Are you adding new handshake operations or why is this required?

Reply:
This is needed for all of the `arith` ops that get created as part of the handshake ops (e.g. `arith.select` for `handshake.mux`). There's a bit of a dilemma here seeing as **all** operations need to be converted/touched in a `handshake.func` - which is done so by `UnitRateConversionPattern` (when no other pattern applies). However, we obviously don't want to run that pattern on these newly created ops since they _do not_ have handshake semantics... suggestions are welcome 🙂!

Reply:
Alright, now I understand the problem. So given the current handshake design, this is indeed necessary.